docs/04-reference/winglang-plugins.md

```diff
@@ -0,0 +1,76 @@
+---
+title: Compiler Plugins
+id: compiler-plugins
+description: Wing Compiler Plugins Reference
+keywords: [Wing reference, Wing language, language, Wing language spec, Wing programming language, compiler plugins]
+---
+
+Wing compiler plugins are used to extend the Wing compiler. They can be used to add non-functional customizations to the compilation's resulting infrastructure. For example, you can use a compiler plugin to enforce a tagging convention on all resources in the resulting infrastructure.
+
+## Plugin Design
+
+A compiler plugin is a JavaScript file that exports multiple functions. These functions are called by the Wing compiler at various points during compilation also referred to as stages.
+
+At each stage the compiler plugin is provided with a context object that allows the plugin to hook into the compilation process and apply customizations.
+
+A single plugin can implement any number of stages. However, it is recommended to only implement the stages that are needed.
+
+## Compilation Stages
+
+Each compilation stage is meant to serve a specific purpose. The following is a list of the stages and their purpose:
+
+### `preSynth` stage
+
+This stage is called before the Wing begins to synthesize. In the context of a terraform-based target like `tf-aws`, this stage will have access to the root node of the construct tree. This allows the plugin to add or change [CDK for Terraform](https://github.com/hashicorp/terraform-cdk) constructs before the tree is synthesized.
+
+The following example adds a bucket to the root node (note this plugin is intended to be used with the `tf-aws` target).
```

Comment:
> note this plugin is intended to be used with the `tf-aws` target.

Feels like we should offer some way for a plugin to validate that it is used within the right context. Let's open an issue to track this need?

Reply:
Yea I have an idea on how we can do this, Ill add it as a suggestion to the issue.

Reply:
https://github.com/winglang/wing/issues/1474